alf/environments/suite_bsuite.py

```diff
@@ -0,0 +1,106 @@
+import alf
+from alf.environments import gym_wrappers, alf_wrappers, alf_gym_wrapper
+from alf.environments.suite_gym import wrap_env
+
+import bsuite
+from bsuite import sweep
+from bsuite.utils import gym_wrapper
+from gym import spaces
+from dm_env import specs
+import numpy as np
+from typing import Any, Dict, Tuple
+
+
+@alf.configurable
+def load(environment=sweep.CARTPOLE_SWINGUP[0],
+         env_id=None,
+         discount=1.0,
+         max_episode_steps=None,
+         gym_env_wrappers=(),
+         alf_env_wrappers=()):
+    """Loads the selected environment and wraps it with the specified wrappers.
+
+    Note that by default a TimeLimit wrapper is used in wrap_env to limit episode 
+    lengths to the default benchmarks defined by the registered environments.
+
+    Args:
+        environment_name (str): Name for the environment to load.
+        env_id (int): (optional) ID of the environment.
+        discount (float): Discount to use for the environment.
+        max_episode_steps (int): If None the max_episode_steps will be set to zero as not
+        all bsuite environments specify max episode lengths.
+            if set to 0 or if there is no max_episode_steps set in the environment's
+            spec.
+        gym_env_wrappers (Iterable): Iterable with references to gym_wrappers
+            classes to use directly on the gym environment.
+        alf_env_wrappers (Iterable): Iterable with references to alf_wrappers
+            classes to use on the ALF environment.
+
+    Returns:
+        An AlfEnvironment instance.
+    """
+
+    env = bsuite.load_from_id(environment)
+    gym_env = BSuiteWrapper(env)
+
+    if max_episode_steps is None:
+        max_episode_steps = 0
```

Comment:
For those envs that do have max_episode_steps, do you need to obtain it from the env so that ALF will correctly set the StepType?

Reply:
It does not seem like the author of bsuite intended others to directly obtain the max_time or max_steps variable, since it is marked with an underscore.

Reply:
> It does not seem like the author of bsuite intended others to directly obtain the max_time or max_steps variable, since it is marked with an underscore.

The reason why we want ALF to differentiate between a normal episode end (success/failure) and a timeout is detailed on this page: https://alf.readthedocs.io/en/latest/notes/knowledge_base.html#about-timestep-discount

In a word, for timeout step, ALF has the step type `LAST` with discount=1, while for normal end step, ALF has the step type `LAST` with discount=0. 

Unfortunately, most envs don't know this subtle difference, including bsuite. For example, in its MountainCar env:

https://github.com/deepmind/bsuite/blob/master/bsuite/environments/mountain_car.py#L89

It treats a success end and timeout the same, which is not correct. 

By obtaining `max_epsiodes` info from the env and pass `max_steps-1` to `suite_gym.wrap_env`, ALF will make sure the correct step type and discount is used. The reason why we should have `max_steps-1` is to prevent bsuite assigning incorrect step type and discount at the last step.

Reply:
Thanks so much for the clarifications! I have added a check for if the environment has instance variable _max_steps, and if it does, set the max_episode_steps to env._max_steps - 1.